### ISSUE TITLE
PCA 初始化时，报错unsupported operand type(s) for -: 'float' and 'NoneType'

### ISSUE THREAD

**Issue by yanglianbing18** [JustGlowing/minisom#191]

**Comment by JustGlowing** [comment]
You get that error because your input data contains null values. Fill them with an imputation technique before training minisom.

**Comment by yanglianbing18** [comment]
检查了数据，没有空值




------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "JustGlowing/minisom"                                                                                    ***@***.***&gt;;
发送时间:&nbsp;2024年7月31日(星期三) 下午3:02
***@***.***&gt;;
***@***.******@***.***&gt;;
主题:&nbsp;Re: [JustGlowing/minisom] PCA 初始化时，报错unsupported operand type(s) for -: 'float' and 'NoneType' (Issue #191)





 
You get that error because your input data contains null values. Fill them with an imputation technique before training minisom.

**Comment by JustGlowing** [comment]
I'm afraid I can't read that. I might offer more help if you post the trace of the error.

**Comment by yanglianbing18** [comment]
Program error is ComplexWarning: Casting complex values to real discards the imaginary part self._weights[i, j] = c1*pc[pc_order[0]] + \

**Comment by yanglianbing18** [comment]
hope you can give me some solution

**Comment by JustGlowing** [comment]
I'm afraid that is only a warning and not the trace of the error you pasted. Make sure that the input data is a numpy array and it contains only scalar values.

**Comment by yanglianbing18** [comment]
But the initial weight printed out is none, and the input data is a numpy array and it contains only scalar values.I try the program , no problem.Here is the code.def pca_weight(data, som, n_rows, n_columns):
&nbsp; &nbsp; pca = PCA(n_components=2)
&nbsp; &nbsp; pca_result = pca.fit_transform(data)
&nbsp; &nbsp; # 计算PCA结果的范围，用于将SOM的网格坐标转换到主成分空间
&nbsp; &nbsp; xmin, xmax = np.min(pca_result[:, 0]), np.max(pca_result[:, 0])
&nbsp; &nbsp; ymin, ymax = np.min(pca_result[:, 1]), np.max(pca_result[:, 1])


&nbsp; &nbsp;# 根据PCA的结果初始化SOM的权重
&nbsp; &nbsp; for i in range(n_rows):
&nbsp; &nbsp; &nbsp; &nbsp; for j in range(n_columns):
&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; # 使用均匀分布在PCA空间内的点来初始化权重
&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; som._weights[i, j, 0] = xmin + (xmax - xmin) * i / (n_rows - 1)
&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; som._weights[i, j, 1] = ymin + (ymax - ymin) * j / (n_columns - 1)
&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; if data.shape[1] &gt; 2:
&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; som._weights[i, j, 2:] = np.mean(data[:, 2:], axis=0)
&nbsp; &nbsp; return som._weights&nbsp; &nbsp; &nbsp;&nbsp;
I hope you can give some suggedtion.Thanks!
.------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "JustGlowing/minisom"                                                                                    ***@***.***&gt;;
发送时间:&nbsp;2024年7月31日(星期三) 晚上9:12
***@***.***&gt;;
***@***.******@***.***&gt;;
主题:&nbsp;Re: [JustGlowing/minisom] PCA 初始化时，报错unsupported operand type(s) for -: 'float' and 'NoneType' (Issue #191)





 
I'm afraid that is only a warning and not the trace of the error you pasted. Make sure that the input data is a numpy array and it contains only scalar values.
